### PR TITLE
QBMM + Preston

### PR DESCRIPTION
**Pull request by anandrdbz**
Verified on CPUs with and without parallel I/o

### CODE REVIEW

**Comment by sbryngelson** [comment]
@anandrdbz please explain your changes to these files in the comments here. For example, I see many goldenfile changes and many others and I'm not sure I understand the provenance of all of them. Please make this easier for me to review by commenting on why and how each file was changed.

**Comment by sbryngelson** [comment]
I do not like the addition of `pb` and `mv` to add of the output routines:
https://github.com/anandrdbz/MFC/blob/21c74df680cc98d8b5f10fb3e8c85dbfa456947a/src/simulation/m_start_up.fpp#L50
for example.

This makes the code more confusing to read. Most folks won't know what these variables are, and they are only used in a very small set of cases. Is there a way to deal with outputting these variables in parallel without adding these as arguments?

File: src/simulation/m_bubbles.fpp [hunk 1]
```diff
@@ -462,7 +464,7 @@ contains
         real(kind(0d0)) :: f_cpbw_KM
 
         if (polytropic) then
-            f_cpbw_KM = Ca*((fR0/fR)**(3.d0*gam)) - Ca + 1d0
+            f_cpbw_KM = (Ca*((fR0/fR)**(3.d0*gam)) - Ca + 1d0)
```
Comment (sbryngelson) [inline src/simulation/m_bubbles.fpp]: why are these parentheses added?

File: src/simulation/m_cbc.fpp [hunk 2]
```diff
@@ -1065,7 +1062,7 @@ contains
 
             if(riemann_solver == 1) then
 !$acc parallel loop collapse(4) gang vector default(present)
-                do i = advxb, advxe
+                do i = 1, advxe
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp]: is this correct? seems weird to me

File: src/simulation/m_cbc.fpp [hunk 3]
```diff
@@ -1219,7 +1217,7 @@ contains
...
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp]: is this correct?!

File: src/simulation/m_cbc.fpp [hunk 4]
```diff
@@ -1303,7 +1302,7 @@ contains
...
                     do r = is3%beg, is3%end
                         do k = is2%beg, is2%end
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp]: where does this come from?

File: src/simulation/m_data_output.fpp [hunk 5]
```diff
@@ -47,9 +47,9 @@ module m_data_output
         !> Write data files
         !! @param q_cons_vf Conservative variables
         !! @param t_step Current time step
-        subroutine s_write_abstract_data_files(q_cons_vf, q_prim_vf, t_step)
+        subroutine s_write_abstract_data_files(q_cons_vf, q_prim_vf, pb, mv, t_step)
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp]: i don't like the addition of variables here.

File: src/simulation/m_data_output.fpp [hunk 6]
```diff
@@ -411,8 +415,8 @@ contains
             if (icfl_max_glb /= icfl_max_glb) then
                 call s_mpi_abort('ICFL is NaN. Exiting ...')
             elseif (icfl_max_glb > 1d0) then
+                !call s_mpi_abort('ICFL is greater than 1.0. Exiting ...')
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp]: why is this commented out?

File: src/simulation/m_data_output.fpp [hunk 7]
```diff
@@ -411,8 +415,8 @@ contains
...
                 print *, 'icfl', icfl_max_glb
-                call s_mpi_abort('ICFL is greater than 1.0. Exiting ...')
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp:415]: why is this deleted?

File: src/simulation/m_data_output.fpp [hunk 8]
```diff
@@ -424,10 +428,13 @@ contains
         !!      conservative variables data files for given time-step.
         !!  @param q_cons_vf Cell-average conservative variables
         !!  @param t_step Current time-step
-    subroutine s_write_serial_data_files(q_cons_vf, q_prim_vf, t_step) ! ---------------------
+    subroutine s_write_serial_data_files(q_cons_vf, q_prim_vf, pb, mv, t_step) ! ---------------------
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp]: i don't like this addition

File: src/simulation/m_data_output.fpp [hunk 9]
```diff
@@ -424,10 +428,13 @@ contains
...
 
         type(scalar_field), dimension(sys_size), intent(IN) :: q_cons_vf
         type(scalar_field), dimension(sys_size), intent(INOUT) :: q_prim_vf
+
+        type(pres_field), intent(INOUT) :: pb, mv
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp]: even if you kept this, shouldn't these be optional variables?

File: src/simulation/m_data_output.fpp [hunk 10]
```diff
@@ -505,6 +512,34 @@ contains
             write (2) q_cons_vf(i)%sf(0:m, 0:n, 0:p); close (2)
         end do
 
+        if(qbmm .and. .not. polytropic) then
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp:509]: please comment your code so it's clear what everything is doing.

File: src/simulation/m_data_output.fpp [hunk 11]
```diff
@@ -528,6 +563,9 @@ contains
             do i = 1, sys_size
                 !$acc update host(q_prim_vf(i)%sf(:,:,:))
             end do
+            if(qbmm) then
+                q_prim_vf(bubxb)%sf = 1d0
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp:562]: what's this doing? comment your code

File: src/simulation/m_data_output.fpp [hunk 12]
```diff
@@ -663,12 +788,14 @@ contains
         !!      conservative variables data files for given time-step.
         !!  @param q_cons_vf Cell-average conservative variables
         !!  @param t_step Current time-step
-    subroutine s_write_parallel_data_files(q_cons_vf, q_prim_vf, t_step) ! --
+    subroutine s_write_parallel_data_files(q_cons_vf, q_prim_vf, pb, mv, t_step) ! --
```
Comment (sbryngelson) [inline src/simulation/m_data_output.fpp]: don't like this addition.

File: src/simulation/m_global_parameters.fpp [hunk 13]
```diff
@@ -343,6 +345,7 @@ contains
         adv_alphan = .false.
         mpp_lim = .false.
         time_stepper = dflt_int
+        weno_vars = dflt_int
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: why was this added? i thought we removed this as an option?

File: src/simulation/m_global_parameters.fpp [hunk 14]
```diff
@@ -585,7 +598,7 @@ contains
                         V0(:) = 1d0
                     else if (nb > 1) then
                         if (R0_type == 1) then
-                            call s_simpson
+                            !call s_simpson
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: i don't understand this, why was this removed?

File: src/simulation/m_global_parameters.fpp [hunk 15]
```diff
@@ -595,15 +608,34 @@ contains
                         stop 'Invalid value of nb'
                     end if
 
-                    print *, 'R0 weights: ', weight(:)
-                    print *, 'R0 abscissas: ', R0(:)
+                    !print *, 'R0 weights: ', weight(:)
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: just delete these lines...

File: src/simulation/m_global_parameters.fpp [hunk 16]
```diff
@@ -595,15 +608,34 @@ contains
...
-                        rhoref = 1.d0
-                        pref = 1.d0
+                    if(.not. qbmm) then
+                        if (polytropic) then
+                            rhoref = 1.d0
+                            pref = 1.d0
+                        end if
                     end if
+
+                    if(qbmm) then
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp:604]: what is this doing? comment your code

File: src/simulation/m_global_parameters.fpp [hunk 17]
```diff
@@ -680,7 +712,7 @@ contains
                     end if
 
                     if (.not. polytropic) then
-                        call s_initialize_nonpoly
+                        !call s_initialize_nonpoly
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: why commented out?

File: src/simulation/m_global_parameters.fpp [hunk 18]
```diff
@@ -816,146 +860,6 @@ contains
 
     end subroutine s_initialize_global_parameters_module ! -----------------
 
-    !> Initializes non-polydisperse bubble modeling
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp:819]: why was this removed?

File: src/simulation/m_mpi_proxy.fpp [hunk 19]
```diff
@@ -96,12 +123,12 @@ contains
 
         #:for VAR in ['t_step_old', 'm', 'n', 'p', 'm_glb', 'n_glb', 'p_glb',  &
             & 't_step_start','t_step_stop','t_step_save','model_eqns',         &
-            & 'num_fluids','time_stepper', 'riemann_solver',  'num_mono',      & 
+            & 'num_fluids','time_stepper', 'weno_vars', 'riemann_solver',      &
```
Comment (sbryngelson) [inline src/simulation/m_mpi_proxy.fpp]: why are you adding weno_vars?!

File: src/simulation/m_mpi_proxy.fpp [hunk 20]
```diff
@@ -1304,10 +1958,10 @@ contains
                         ! Send/receive buffer to/from bc_x%end/bc_x%beg
                         call MPI_SENDRECV( &
                             q_cons_buff_send(0), &
-                            buff_size*sys_size*(m + 2*buff_size + 1)*(n + 2*buff_size + 1), &
+                            buff_size*v_size*(m + 2*buff_size + 1)*(n + 2*buff_size + 1), &
```
Comment (sbryngelson) [inline src/simulation/m_mpi_proxy.fpp:1961]: what's the difference between v_size and sys_size?

File: src/simulation/m_qbmm.fpp [hunk 21]
```diff
@@ -205,115 +513,210 @@ contains
...
         type(int_bounds_info), intent(IN) :: ix, iy, iz
 
-        real(kind(0d0)), dimension(nmom) :: moms
-        real(kind(0d0)), dimension(nb) :: Rvec
-        real(kind(0d0)), dimension(nnode, nb) :: wght, abscX, abscY
+        real(kind(0d0)), dimension(nmom) :: moms, msum
+        real(kind(0d0)), dimension(nnode, nb) :: wght, abscX, abscY, wght_pb, wght_mv, wght_ht , ht 
         real(kind(0d0)), dimension(nterms, 0:2, 0:2) :: mom3d_terms, coeff
-        real(kind(0d0)) :: pres, rho, nbub, c, alf, R3, momsum
+        real(kind(0d0)) :: pres, rho, nbub, c, alf, R3, momsum, drdt, drdt2, chi_vw, x_vw, rho_mw, k_mw, T_bar, grad_T
```
Comment (sbryngelson) [inline src/simulation/m_qbmm.fpp:555]: please add comments explaining what the variables are in doxygen format

File: src/simulation/m_qbmm.fpp [hunk 22]
```diff
@@ -205,115 +513,210 @@ contains
...
+                            do r = 2, nmom                                                                     
+                                moms(r) = q_prim_vf(bubmoms(q, r))%sf(id1, id2, id3) 
                             end do
 
-                           
+                            moms(1) = 1d0
 
                             call s_chyqmom(moms, wght(:, q), abscX(:, q), abscY(:, q))
 
+                            if(polytropic) then
```
Comment (sbryngelson) [inline src/simulation/m_qbmm.fpp:610]: what is this doing? comment your code

File: src/common/include/inline_conversions.fpp [hunk 23]
```diff
@@ -30,15 +30,19 @@
         elseif (((model_eqns == 4) .or. (model_eqns == 2 .and. bubbles))) then
             ! Sound speed for bubble mmixture to order O(\alpha)
 
+            n_tait = (1d0/gamma + 1d0)
```
Comment (sbryngelson) [inline src/common/include/inline_conversions.fpp]: why do we have n_tait, B_tait, pi_inf, _and_ gamma?

File: src/common/m_mpi_common.fpp [hunk 24]
```diff
@@ -62,27 +62,36 @@ contains
     end subroutine s_mpi_initialize ! --------------------------------------
 
 
-    subroutine s_initialize_mpi_data(q_cons_vf) ! --------------------------
+    subroutine s_initialize_mpi_data(q_cons_vf, pb, mv) ! --------------------------
```
Comment (sbryngelson) [inline src/common/m_mpi_common.fpp]: don't like the addition of these variables - makes the code very confusing

File: src/post_process/m_data_output.f90 [hunk 25]
```diff
@@ -939,6 +1029,49 @@ subroutine s_write_variable_to_formatted_database_file(varname, t_step)
 
     end subroutine s_write_variable_to_formatted_database_file ! -----------
```
Comment (sbryngelson) [inline src/post_process/m_data_output.f90]: I think @wilfonba removed this feature intentionally. please do not add it back.

File: src/post_process/m_data_output.f90 [hunk 26]
```diff
@@ -973,9 +1106,14 @@ subroutine s_finalize_data_output_module() ! -------------------------
         ! that were written to the formatted database file(s). Note that the
         ! root variable is only deallocated in the case of a 1D computation.
         deallocate (q_sf)
+        if (coarsen_silo) deallocate (coarse_x_q_sf, coarse_xy_q_sf, coarse_xyz_q_sf)
         if (n == 0) deallocate (q_root_sf)
         if (grid_geometry == 3) then
-            deallocate (cyl_q_sf)
+            if (coarsen_silo) then
```
Comment (sbryngelson) [inline src/post_process/m_data_output.f90]: @wilfonba removed this feature intentionally (it is fairly useless). Please do not add it back.

File: src/pre_process/m_assign_variables.f90 [hunk 27]
```diff
@@ -209,6 +210,83 @@ subroutine s_assign_patch_mixture_primitive_variables(patch_id, j, k, l, &
 
     end subroutine s_assign_patch_mixture_primitive_variables ! ------------
 
+    subroutine s_perturb_primitive(patch_id, j, k, l, q_prim_vf)
```
Comment (sbryngelson) [inline src/pre_process/m_assign_variables.f90]: what is this doing?

File: src/pre_process/m_data_output.fpp [hunk 28]
```diff
@@ -253,20 +342,55 @@ contains
...
+            end if
         end if
 
     end subroutine s_write_serial_data_files ! ------------------------------------
 
     !> Writes grid and initial condition data files in parallel to the "0"
         !!  time-step directory in the local processor rank folder
         !! @param q_cons_vf The conservative variables
-    subroutine s_write_parallel_data_files(q_cons_vf) ! --
+    subroutine s_write_parallel_data_files(q_cons_vf, pb, mv) ! --
```
Comment (sbryngelson) [inline src/pre_process/m_data_output.fpp]: don't like this addition

File: src/pre_process/m_start_up.fpp [hunk 29]
```diff
@@ -49,15 +49,20 @@ module m_start_up
 
         end subroutine s_read_abstract_grid_data_files ! ---------------
 
-        subroutine s_read_abstract_ic_data_files(q_cons_vf) ! -----------
+        subroutine s_read_abstract_ic_data_files(q_cons_vf, pb, mv) ! -----------
```
Comment (sbryngelson) [inline src/pre_process/m_start_up.fpp]: don't like this addition!

File: src/pre_process/m_start_up.fpp [hunk 30]
```diff
@@ -358,12 +363,14 @@ contains
         !!      the pre-process as a starting point in the creation of an
         !!      all new initial condition.
         !! @param q_cons_vf Conservative variables
-    subroutine s_read_serial_ic_data_files(q_cons_vf) ! ---------------------------
+    subroutine s_read_serial_ic_data_files(q_cons_vf, pb, mv) ! ---------------------------
```
Comment (sbryngelson) [inline src/pre_process/m_start_up.fpp]: don't like this addition

File: src/pre_process/p_main.f90 [hunk 31]
```diff
@@ -120,11 +131,11 @@ program p_main
         end if
     end if
 
-    if (old_ic) call s_read_ic_data_files(q_cons_vf)
+    if (old_ic) call s_read_ic_data_files(q_cons_vf, pb, mv)
```
Comment (sbryngelson) [inline src/pre_process/p_main.f90]: don't like this addition

File: src/simulation/m_bubbles.fpp (hunk 1 above)
Comment (anandrdbz) [reply src/simulation/m_bubbles.fpp]: Not necessary, I'll remove these

File: src/simulation/m_cbc.fpp (hunk 4 above)
Comment (anandrdbz) [reply src/simulation/m_cbc.fpp]: I believe these are incorrect as well, I was certain these had 1, sys_size at first which was giving a bug, which I then corrected to 1, advxe. But it really should be advxb, advxe (it was perhaps corrected at some point, but the merge with my branch messed it up)

I'll fix these in the new commit

File: src/pre_process/p_main.f90 (hunk 31 above)
Comment (anandrdbz) [reply src/pre_process/p_main.f90]: I could make these variables global in both simulation and pre_process, which would no longer require their presence in the function arguments

Comment (sbryngelson) [reply src/pre_process/p_main.f90]: that might be a good option

File: src/pre_process/m_assign_variables.f90 (hunk 27 above)
Comment (anandrdbz) [reply src/pre_process/m_assign_variables.f90]: This was included to emulate the 1D perturbation for a stable shock (using shock jump relations) which was subsequently used for FFT analysis . Currently it's not being used, I thought I'd add it in case we end up using it again

**Comment by anandrdbz** [comment]
> please (very briefly) explain your changes to these files in the comments here. For example, I see many goldenfile changes and many others and I'm not sure I understand the provenance of all of them. Please make this easier for me to review by commenting on why and how each file was changed.

Golden file changes are for all the qbmm tests (including previous polytropic ones). These are required due to a fundamental mistake in how we were dealing with moments, where the averaging wasn't being done properly in the R and Rdot directions.

File: src/pre_process/m_assign_variables.f90 (hunk 27 above)
Comment (sbryngelson) [reply src/pre_process/m_assign_variables.f90]: It could be useful to add/keep. Please just add code comments for it. Also it looks like `patch_id` is not used, so you can remove that argument. You might also add an `example/` case that uses it (non-hard-coded).

**Comment by sbryngelson** [comment]
> > please (very briefly) explain your changes to these files in the comments here. For example, I see many goldenfile changes and many others and I'm not sure I understand the provenance of all of them. Please make this easier for me to review by commenting on why and how each file was changed.
> 
> Golden file changes are for all the qbmm tests (including previous polytropic ones). These are required due to a fundamental mistake in how we were dealing with moments, where the averaging wasn't being done properly in the R and Rdot directions.

Got it, I agree.

**Comment by sbryngelson** [comment]
> @sbryngelson , I have also removed the redundant s_assign_patch_bubbles which was mentioned in one of the other issues

For reference https://github.com/MFlowCode/MFC/issues/177 task 2

**Comment by sbryngelson** [comment]
This change appears to have mistakes: https://github.com/MFlowCode/MFC/pull/180/files#diff-c46f7cd6ce1da4d2b4e71a99cd0b6a96e70bb49fa0c73679edd804dcd7f81c40

In particular, @wilfonba removed `coarsen_silo` intentionally and you are putting it back in. Please fix.

File: src/post_process/m_global_parameters.f90 [hunk 32]
```diff
@@ -50,7 +50,7 @@ module m_global_parameters
     !> @name Cell-boundary locations in the x-, y- and z-coordinate directions
     !> @{
     real(kind(0d0)), allocatable, dimension(:) :: x_cb, x_root_cb, y_cb, z_cb
-    real(kind(0.0)), allocatable, dimension(:) :: x_cb_s, y_cb_s, z_cb_s
+    real(kind(0d0)), allocatable, dimension(:) :: coarse_x_cb, coarse_y_cb, coarse_z_cb
```
Comment (sbryngelson) [inline src/post_process/m_global_parameters.f90]: remove

File: src/post_process/m_global_parameters.f90 [hunk 33]
```diff
@@ -301,11 +318,13 @@ subroutine s_assign_default_values_to_user_inputs() ! ------------------
         cons_vars_wrt = .false.
         c_wrt = .false.
         omega_wrt = .false.
-		qm_wrt = .false.
+        qm_wrt = .false.
         schlieren_wrt = .false.
 
         schlieren_alpha = dflt_real
 
+        fourier_decomp = .false.
```
Comment (sbryngelson) [inline src/post_process/m_global_parameters.f90]: why is this added?

File: src/post_process/m_mpi_proxy.fpp [hunk 34]
```diff
@@ -163,9 +163,10 @@ contains
 
         #:for VAR in [ 'cyl_coord', 'adv_alphan', 'mpp_lim', 'mixture_err',    &
             & 'alt_soundspeed', 'hypoelasticity', 'parallel_io', 'rho_wrt',    &
-            & 'E_wrt', 'pres_wrt', 'gamma_wrt', 'polytropic', 'polydisperse',  & 
+            & 'E_wrt', 'pres_wrt', 'gamma_wrt',                & 
             & 'heat_ratio_wrt', 'pi_inf_wrt', 'pres_inf_wrt', 'cons_vars_wrt', & 
-            & 'prim_vars_wrt', 'c_wrt', 'qm_wrt','schlieren_wrt', 'bubbles']
+            & 'prim_vars_wrt', 'c_wrt', 'qm_wrt','schlieren_wrt', 'bubbles',   &
+            & 'polytropic', 'fourier_decomp', 'polydisperse' ]
```
Comment (sbryngelson) [inline src/post_process/m_mpi_proxy.fpp]: it looks like you added `fourier_decomp` here why?

File: src/post_process/m_start_up.f90 [hunk 35]
```diff
@@ -53,7 +53,8 @@ subroutine s_read_input_file() ! ---------------------------------------
             omega_wrt, qm_wrt, schlieren_wrt, schlieren_alpha, &
             fd_order, mixture_err, alt_soundspeed, &
             flux_lim, flux_wrt, cyl_coord, &
-            parallel_io, rhoref, pref, bubbles, R0ref, nb, &
+            parallel_io, coarsen_silo, fourier_decomp, &
```
Comment (sbryngelson) [inline src/post_process/m_start_up.f90]: remove new variables why are coarsen silo and fourier decomp here?!

**Review by sbryngelson — CHANGES_REQUESTED** [review]

File: src/simulation/m_global_parameters.fpp (hunk 18 above)
Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp:819]: Moved to common directory

**Comment by anandrdbz** [comment]
@sbryngelson I have fixed all the files individually which had merge issues (it added back coarsen_silo inadvertently). They should all be resolved now

**Comment by sbryngelson** [comment]
@anandrdbz this looks good. Can you do one (I promise) final thing, check the run times for the GPU code vs CPU code with the different bubble models and see how they differ before and after your PR? 

I mostly just want to make sure that the `QBMM without Preston` and the `method of classes` cases are just as fast as they used to be.

You can run the cases on Summit or some similar machine and report the differences here in the PR comments.

**Review by sbryngelson — APPROVED** [review]

**Comment by anandrdbz** [comment]
@sbryngelson So there was a minor gpu allocation bug for large cases that is now fixed for qbmm + polytropic 

Here are the results for slowdown/speedup on Summit for a 3D problem with 1M points : 

CPU + Method of classes : No slowdown 
CPU + QBMM w/ Poly : 1% slowdown

GPU + Method of classes : No slowdown
GPU + QBMM w/ Poly : 15% slowdown

MOC shows no slowdown as this part of the code was not touched 

QBMM w/ Polytropic does show minor slowdown (esp on GPUs) due to introduction of terms for Re, Surf tension and pv (total of 26 terms in bubble dynamics as opposed to only 11 previously). All of this was missing in the previous implementation of qbmm so it's not a fair comparison. However, the slowdown is quite minor (15%) in the grand scheme of things

**Comment by anandrdbz** [comment]
Also, I noticed the GPU test on phoenix failed, and I believe this is due to issues on phoenix. All the tests pass on Bridges2 and Summit, but I am unable to get a compute node on phoenix. In fact the case that failed was the very first one which has nothing to do with qbmm (or the change that I just made) . 

@wilfonba was having similar issues with GPUs on phoenix today as well
